Composer/packages/extensions/obiformeditor/src/Form/widgets/LuEditorWidget.tsx

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import React, { useState, useMemo, useEffect } from 'react';
+import { LuEditor } from '@bfc/code-editor';
+import debounce from 'lodash/debounce';
+import { LuIntentSection } from '@bfc/shared';
+import { LuFile, filterSectionDiagnostics } from '@bfc/indexers';
+
+import { FormContext } from '../types';
+
+interface LuEditorWidgetProps {
+  formContext: FormContext;
+  name: string;
+  height?: number | string;
+  onChange: (template?: string) => void;
+}
+
+export const LuEditorWidget: React.FC<LuEditorWidgetProps> = props => {
+  const { formContext, name, height = 250 } = props;
+  const luFileId = formContext.currentDialog.id;
+  const luFile: LuFile | null = formContext.luFiles.find(f => f.id === luFileId);
```

Comment:
This component has several problems. The usage of `useEffect` hook is anti-pattern meanwhile the data flow is not well designed (as an example, the `onChange` handler submits data to both local states and external API).

The props and local states are mixed together, makes your component hard to maintain.

You really need to rethink what's your scenario is and carefully define the shape of your input props to make sure you can fully leverage change detection mechanism.

Before your modification on this file, do read these 2 docs:
1. https://reactjs.org/docs/react-component.html#unsafe_componentwillreceiveprops
1. https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#preferred-solutions

Reply:
I know state management of Shell has historical problems, consider using a Class Component with lifecycle methods before Leilei's fix on Shell states.